course_contents/2_intro_to_python/notes/02. Loops.md

```diff
@@ -0,0 +1,159 @@
+### Loops
+- loops are used to run blocks of code multiple times
+- there are two kinds of loops in Python:
+    1. while loops
+    2. for loops
+
+**Extra resources**
+[What is a loop?](https://python.tecladocode.com/3_continuous_countries/1_loops.html#what-is-a-loop)
+
+### The `while` loop
+- while loop runs an indefinite number of times i.e. while loop runs until the given conditional becomes False
+- syntax for writing the while loop:
+    1. use the `while` keyword at the beginning of the loop
+    2. the keyword should be followed with a correctly written conditional with a colon at the end of it
+    3. indented block of code that will run until the loop breaks
+
+An example of the properly written `while` loop:
+
+```py
+counter = 10
+while counter > 0:
+  print(f"{counter} seconds left until New Year!")
+  counter -= 1
+print("Happy New Year!")
+```
+
+An example of the `while` loop that will run until the program is manually stopped:
+
+```py
+counter = 10
+while counter > 0: # the counter never gets changed, causing the conditional to be always True
+  print(f"{counter} seconds left until New Year!")
+```
+
+### The `for` loop
+- for loop runs a given number of times
+- general syntax for `for` loop is:
+
+```py
+for variable in iterable:
+    block_of_code
+```
+- an iterable is an object that can be iterated over e.g. a list
+- an object is iterable if it can return the iterator object (more on that in the OOP section)
+
+An example of the `for` loop:
+
+```py
+numbers = [1,2,3,4,5]
+for i in numbers:
+    print(i)
+```
+
+**Extra resources**
+[How does the for loop work in Python?](https://python.tecladocode.com/3_continuous_countries/1_loops.html#how-does-the-for-loop-work-in-python)
+
+#### The `range()` method
+- the `range()` method returns a sequence of integers in a given range
+- arguments that are passed to the method are:
+    1. `start` - an integer marking the start of a sequence
+    2. `stop` - an integer marking the end of a sequence. Important - `range()` returns integers up to `stop-1`
+    3. `step` - an integer representing the size of the step i.e. the difference between consecutive integers in a sequence
+- the numbers are generated by starting from `start` and incrementing it by `step` until `stop-1` is reached
+- the only argument that has to be stated explicitly is the `stop`
+- if not specified, Python will implicitly set `start` to 0 and `step` to 1
+
+An example of the `range()` method used in a `for` loop:
+
+```py
+for i in range(5): #the range() method can be used to specify number of iterations of a loop
+  print(i) #the numbers printed will be: 0,1,2,3,4
+```
+
+### The `break` statement
+- the `break` statement is used to stop the loop from running when an external condition is reached
```

Comment:
I wouldn't be so sure about this. If the condition of a while loop changes, the loop will stop. Can you think of a practical example where we would really use `break` and reword this?

Reply:
^ I like this suggestion, because it's not necessary that `break` is called as a result of an external condition (although that is normally how it's used).

Reply:
Changed it, although I formulated it in a different way:

> the `break` statement is used to stop the loop from iterating

If you feel like Vlad's suggestion sounds better we can go with it

Reply:
I think Vlad's suggestion here might be better, since your wording suggests that there will be 0 iterations. In reality, there will be as many iterations as happen until the `break` is encountered 🤔